src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionCommandHandler.cs

```diff
@@ -0,0 +1,156 @@
+
+        internal AbstractCommentSelectionCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            Contract.ThrowIfNull(undoHistoryRegistry);
+            Contract.ThrowIfNull(editorOperationsFactoryService);
+
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        protected static VSCommanding.CommandState GetCommandState(ITextBuffer buffer)
+        {
+            if (!buffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return VSCommanding.CommandState.Unspecified;
+            }
+
+            return VSCommanding.CommandState.Available;
+        }
+
+        protected static void Format(ICommentSelectionService service, ITextSnapshot snapshot, IEnumerable<CommentTrackingSpan> changes, CancellationToken cancellationToken)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return;
+            }
+
+            // Only format uncomment actions.
+            var textSpans = changes
+                .Where(change => change.Operation == Operation.Uncomment)
+                .Select(uncommentChange => uncommentChange.ToSnapshotSpan(snapshot).Span.ToTextSpan())
+                .ToImmutableArray();
+            var newDocument = service.FormatAsync(document, textSpans, cancellationToken).WaitAndGetResult(cancellationToken);
+            newDocument.Project.Solution.Workspace.ApplyDocumentChanges(newDocument, cancellationToken);
+        }
+
+        protected static void InsertText(List<TextChange> textChanges, int position, string text)
+        {
+            textChanges.Add(new TextChange(new TextSpan(position, 0), text));
+        }
+
+        protected static void DeleteText(List<TextChange> textChanges, TextSpan span)
+        {
+            textChanges.Add(new TextChange(span, string.Empty));
+        }
+
+        internal bool ExecuteCommand(ITextView textView, ITextBuffer subjectBuffer, Operation operation, CommandExecutionContext context)
+        {
+            var title = GetTitle(operation);
+            var message = GetMessage(operation);
+
+            using (context.OperationContext.AddScope(allowCancellation: false, message))
+            {
+
+                var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = GetService(document);
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var trackingSpans = new List<CommentTrackingSpan>();
+                var textChanges = new List<TextChange>();
```

Comment:
1. why do we need to define tehse here, instead of having CollectEdits just return this information?
2. can we explain what the tracking spans are for?

Reply:
1.  Could return it.  I'm open to either, so I just did it the same as it was in the existing comment handler.  If return is preferred I can do that.
2.  Sets the spans to format and spans to select once the operation is completed.

Reply:
1. please have it just return it.
2. sorry to be unclear.  when i say things like "can we explain" it generally means: can we have a comment in the code explaining this so that anyone in the future will understand it.

Thanks! :)

Reply:
Got it, will change.